client_code/animation.py

```diff
@@ -0,0 +1,542 @@
+
+    def persist(self) -> None:
+        "Explicitly persists an animation, when it would otherwise be removed."
+
+    def reverse(self) -> None:
+        "Reverses playback direction and plays"
+
+    def updatePlaybackRate(self, playback_rate) -> None:
+        "The new speed to set. A positive number (to speed up or slow down the animation), a negative number (to reverse), or zero (to pause)."
+
+    def wait(self) -> None:
+        "Animations are not blocking. Call the wait function to wait for an animation to finish in a blocking way"
+
+    @property
+    def playbackRate(self) -> int or float:
+        "gets or sets the playback rate"
+
+    @property
+    def onfinish(self):
+        "set a callback for when the animation finishes"
+
+    @property
+    def oncancel(self):
+        "set a callback for when the animation is cancelled"
+
+    @property
+    def onremove(self):
+        "set a callback for when the animation is removed"
+
+
+def _animate(component, keyframes, options, use_ghost=False):
+    if isinstance(keyframes, Transition):
+        keyframes = keyframes._compute()
+    el = _dom_node(component)
+
+    if use_ghost:
+        _animate_ghost(el, keyframes, options)
+
+    return Animation(_a=el.animate(keyframes, options))
+
+
+_window.Function(
+    "_animate",
+    """
+KeyframeEffect.prototype.animate = function(component, ghost=false) {
+    const keyframes = this.getKeyframes();
+    const timing = this.getTiming();
+    return _animate(component, keyframes, timing, ghost);
+}
+""",
+)(_animate)
+
+
+class Effect:
+    """Create an effect that can later be used to animate a component.
+    The first argument should be a Transition object.
+    The remainder of the values are timing options"""
+
+    def __new__(cls, transition=None, **timings):
+        if isinstance(transition, Transition):
+            transition = transition._compute()
+        return _window.KeyframeEffect(None, transition, timings)
+
+    def __init__(
+        self,
+        transition=None,
+        *,
+        duration=333,
+        delay=0,
+        direction="normal",
+        easing="linear",
+        endDelay=0,
+        fill="none",
+        iterations=1,
+        iterationStart=0,
+        composite="replace",
+    ):
+        pass
+
+    def animate(self, component, use_ghost=False) -> Animation:
+        """animate a component using an effect.
+        If use_ghost is True a ghost element will be animated.
+        use_ghoste allows components to be animated outside of their container
+        """
+        return Animation(_a=None)  # just so the autocompleter knows the return type
+
+    def getKeyframes(self, component):
+        "Returns the computed keyframes that make up this effect"
+
+    def getTiming(self, component):
+        "The EffectTiming object associated with the animation"
+
+
+def animate(
+    component,
+    transition=None,
+    *,
+    duration=333,
+    start_at=None,
+    end_at=None,
+    use_ghost=False,
```

Comment:
Three variable names that I'm not convinced about.
These can be what ever we like because they are custom.

docs for these three variables:

>     `use_ghost`: when set to `True`, will animate a ghost element (i.e. a visual copy).
>     Using a ghost element will allow the component to be animated outside of its container
> 
>     `start_at`, `end_at`: Can be set to a `Component` or `DOMRect` (i.e. a computed position of a component from `get_bounding_rect`)
>     If either `start_at` or `end_at` are set this will determine the start/end position of the animation
>     If one value is set and the other omitted the omitted value will be assumed to be the current position of the component.
>     A ghost element is always used when `start_at` / `end_at` are set.